### PR TITLE
[SHELL32] Implement CSIDL_CONNECTIONS and fix CSIDL_PRINTERS

### PR DESCRIPTION
**Pull request by whindsaks**
CSIDL_CONNECTIONS is needed by external callers, mainly NetShell PR #6905

### CODE REVIEW

File: dll/win32/shell32/utils.cpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ SHBindToObjectEx(
     return hr;
 }
 
-HRESULT SHBindToObject(
+EXTERN_C HRESULT SHBindToObject(
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:205]:
```suggestion
EXTERN_C
HRESULT SHBindToObject(
```

File: dll/win32/shell32/precomp.h [hunk 2]
```diff
@@ -181,7 +181,7 @@ Shell_DisplayNameOf(
     _Out_ LPWSTR pszBuf,
     _In_ UINT cchBuf);
 
-HRESULT SHBindToObject(
+EXTERN_C HRESULT SHBindToObject(
```
Comment (HBelusca) [inline dll/win32/shell32/precomp.h:184]:
```suggestion
EXTERN_C
HRESULT SHBindToObject(
```

File: dll/win32/shell32/wine/shellpath.c [hunk 3]
```diff
@@ -3099,6 +3099,32 @@ BOOL WINAPI SHGetSpecialFolderPathW (
                             szPath) == S_OK;
 }
 
+#ifdef __REACTOS__
+HRESULT WINAPI SHGetFolderLocationHelper(HWND hwnd, int nFolder, REFCLSID clsid, LPITEMIDLIST *ppidl)
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:3103]: Since it's a non-exported internal helper, better remove the "WINAPI" thing.

**Review by HBelusca — APPROVED** [review]
Minor formatting stuff but otherwise LGTM

**Review by learn-more — APPROVED** [review]

File: dll/win32/shell32/wine/shellpath.c [hunk 4]
```diff
@@ -3099,6 +3099,32 @@ BOOL WINAPI SHGetSpecialFolderPathW (
...
+{
+    HRESULT hr;
+    IShellFolder *psf;
+    LPITEMIDLIST parent, child;
+    EXTERN_C HRESULT SHBindToObject(IShellFolder *psf, LPCITEMIDLIST pidl, REFIID riid, void **ppvObj);
```
Comment (katahiromz) [inline dll/win32/shell32/wine/shellpath.c:3108]: It's better to declare function in appreciate place.

File: dll/win32/shell32/precomp.h (hunk 2 above)
Comment (whindsaks) [reply dll/win32/shell32/precomp.h:184]: The formatting in this file is all over the place, even SHBindToObject vs SHBindToObjectEx are different!

File: dll/win32/shell32/wine/shellpath.c (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shell32/wine/shellpath.c:3108]: What is the proper place? It was **you** that placed the declaration for this function in precomp.h

File: dll/win32/shell32/wine/pidl.c [hunk 5]
```diff
@@ -1680,6 +1680,13 @@ LPITEMIDLIST _ILCreateControlPanel(void)
 
 LPITEMIDLIST _ILCreatePrinters(void)
 {
+#ifdef __REACTOS__
+    // Note: Wine returns the PIDL as it was in Windows 95, NT5 moved it into CSIDL_CONTROLS
+    extern HRESULT WINAPI SHGetFolderLocationHelper(HWND hwnd, int nFolder, REFCLSID clsid, LPITEMIDLIST *ppidl);
```
[the remark this answers is not in the document]
Comment (whindsaks) [reply dll/win32/shell32/wine/pidl.c]: Nobody except SHGetFolderLocation is really supposed to call this, just doing this to fix Wine brokenness.

**Comment by binarymaster** [comment]
Please associate with JIRA issue: [CORE-19562](https://jira.reactos.org/browse/CORE-19562) when merging.

**Comment by binarymaster** [comment]
I guess it's too late...

**Comment by whindsaks** [comment]
> I guess it's too late...

Yeah, sorry. CORE-19562 is more about the action taken by the button and not really the shell folders being broken.

**Pull request merged**